toolchains/tool.bzl

```diff
@@ -0,0 +1,115 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Implementation of tool"""
+
+load("//toolchains/private:collect.bzl", "collect_data", "collect_provider")
+load(
+    ":toolchain_info.bzl",
+    "ToolCapabilityInfo",
+    "ToolInfo",
+)
+
+def _tool_impl(ctx):
+    exe_info = ctx.attr.src[DefaultInfo]
+    if exe_info.files_to_run != None and exe_info.files_to_run.executable != None:
+        exe = exe_info.files_to_run.executable
```

Comment:
Could we keep and stage the entire `files_to_run`? Runfiles are a pretty powerful mechanism and also lend themselves to optimizations (Merkle tree computation, symlinking entire runfiles dirs instead of the individual files into sandboxes, ...) that input files won't support.

Reply:
That was my intention, and should be happening in `collect_data` in `collect.bzl`, but I'm not particularly familiar with how files_to_run works, so I may be doing it wrong.

Data is currently a `List[Either[Depset[File], FilesToRunProvider]]`. Is that correct?
See `collect_data` for the implementation, which does attempt to collect the entire `files_to_run`.